src/core/map.c

```diff
@@ -0,0 +1,522 @@
+                return 0;
+        }
+
+#ifdef USE_POSIX
+        if (pthread_mutex_lock(&map->mutex) != 0) {
+                return -1;
+        }
+#endif
+        return 0;
+}
+
+// Attempt to unlock the map
+static int map_unlock(struct map *map)
+{
+        assert(map);
+
+        if (map->setting_flags & MAP_FLAG_NO_LOCKING) {
+                return 0;
+        }
+
+#ifdef USE_POSIX
+        if (pthread_mutex_unlock(&map->mutex) != 0) {
+                return -1;
+        }
+#endif
+        return 0;
+}
+
+// Use the robin hood hashing method to probe the map for a sutable location to
+// place the provided entry
+static enum map_error map_probe(struct map *map, struct map_entry *entry,
+                                uint32_t initial_index)
+{
+        // As we probe through the map these will continually get updated
+        uint32_t probe_position = initial_index;
+        struct map_entry *insert_entry = entry;
+
+        for (;;) {
+                if (probe_position == map->length) {
+                        probe_position = 0;
+                }
+
+                assert(probe_position < map->length);
+
+                struct map_entry *current_entry = map->entries[probe_position];
+
+                if (!is_entry_present(current_entry)) {
+                        // Insert here
+                        map->entries[probe_position] = insert_entry;
+
+                        return NO_ERROR;
+                }
+
+                uint32_t insert_dib
+                        = get_dib(map, insert_entry, probe_position);
+
+                uint32_t current_dib
+                        = get_dib(map, current_entry, probe_position);
+
+                // If the current entry has a lower dib then the entry we are
+                // trying to insert then we swap the entries
+                if (current_dib < insert_dib) {
+                        map->entries[probe_position] = insert_entry;
+
+                        insert_entry = current_entry;
+                        probe_position += 1;
+
+                        continue;
+                }
+
+                // Keep probing
+                probe_position += 1;
+        }
+
+        // Should never happen
+        return GENERIC_ERROR;
+}
+
+static enum map_error map_search(const struct map *map, const char *key,
+                                 uint32_t *index)
+{
+        if (key == NULL) {
+                return GENERIC_ERROR;
+        }
+
+        // Initialize index with initial bucket index
+        uint32_t current_index = map->hash_function(key) % map->length;
+
+        for (size_t i = 0; i <= map->length; ++i) {
+                current_index += (uint32_t)i;
+
+                if ((current_index + i) >= map->length) {
+                        current_index = 0;
+                }
+
+                struct map_entry *entry = map->entries[current_index];
+
+                if (!is_entry_present(entry) || strcmp(key, entry->key) != 0) {
```

Comment:
if entry is not present it will still fall through to the second part of the conditional check making `entry->key` UB

Reply:
nevermind - stupid